tests/test_optim.py

```diff
@@ -83,47 +83,47 @@ def test_minimize(tensorlib, precision, optimizer, do_grad, do_stitch):
             # no grad, minuit, 32b - not very consistent for pytorch
             'no_grad-minuit-numpy-32b': [0.49622172117233276, 1.0007264614105225],
             #    nb: macos gives different numerics than CI
-            # 'no_grad-minuit-pytorch-32b': [0.7465415000915527, 0.8796938061714172],
             'no_grad-minuit-pytorch-32b': [0.9684963226318359, 0.9171305894851685],
             'no_grad-minuit-tensorflow-32b': [0.5284154415130615, 0.9911751747131348],
-            # 'no_grad-minuit-jax-32b': [0.5144518613815308, 0.9927923679351807],
             'no_grad-minuit-jax-32b': [0.49620240926742554, 1.0018986463546753],
             # no grad, minuit, 64b - quite consistent
             'no_grad-minuit-numpy-64b': [0.5000493563629738, 1.0000043833598724],
             'no_grad-minuit-pytorch-64b': [0.5000493563758468, 1.0000043833508256],
             'no_grad-minuit-tensorflow-64b': [0.5000493563645547, 1.0000043833598657],
             'no_grad-minuit-jax-64b': [0.5000493563528641, 1.0000043833614634],
             # do grad, minuit, 32b
-            'do_grad-minuit-pytorch-32b': [0.5017611384391785, 0.9997190237045288],
-            'do_grad-minuit-tensorflow-32b': [0.5012885928153992, 1.0000673532485962],
-            # 'do_grad-minuit-jax-32b': [0.5029529333114624, 0.9991086721420288],
-            'do_grad-minuit-jax-32b': [0.5007095336914062, 0.9999282360076904],
+            # large divergence by tensorflow and pytorch
+            'do_grad-minuit-pytorch-32b': [0.9731879234313965, 0.9999999403953552],
+            'do_grad-minuit-tensorflow-32b': [0.9366918206214905, 0.9126002788543701],
+            'do_grad-minuit-jax-32b': [0.5003563165664673, 0.9998618364334106],
```

Comment:
It isn't really clear to me what is happenng that there is such a drastic shift here.

Reply:
Could it be that the PyTorch/TensorFlow fail to converge?

Reply:
@alexander-held That's a good question, but the are succeeding 

```python
import pyhf

precision = "32b"
do_grad = True
do_stitch = True
strategy = None
# strategy = 2
verbose = 1

names = ["TensorFlow", "PyTorch", "JAX"]
backends = [
    pyhf.tensor.tensorflow_backend,
    pyhf.tensor.pytorch_backend,
    pyhf.tensor.jax_backend,
]
for name, backend in zip(names, backends):
    print(f"\n{name}\n")
    pyhf.set_backend(
        backend(precision=precision),
        pyhf.optimize.minuit_optimizer(verbose=verbose, strategy=strategy),
    )
    model = pyhf.simplemodels.hepdata_like([50.0], [100.0], [10.0])
    data = pyhf.tensorlib.astensor([125.0] + model.config.auxdata)
    result = pyhf.infer.mle.fit(
        data, model, do_grad=do_grad, do_stitch=do_stitch, return_result_obj=True
    )
    fit_params = result[0]
    fit_result = result[1]
    print(f"fit params: {[pyhf.tensorlib.tolist(x) for x in fit_params]}")
    print(fit_result.minuit)
    print(f"success: {fit_result.success}")
```
gives

```
TensorFlow

W MnPosDef non-positive diagonal element in covariance matrix[ 0 ] = -1.49124e+08
W MnPosDef non-positive diagonal element in covariance matrix[ 1 ] = -4983.71
W MnPosDef Added to diagonal of Error matrix a value 1.49124e+08
fit params: [0.9366918206214905, 0.9126002788543701]
┌──────────────────────────────────┬──────────────────────────────────────┐
│ FCN = 15.21                      │              Nfcn = 74               │
│ EDM = 1.14e+03 (Goal: 0.0002)    │              Ngrad = 1               │
├───────────────┬──────────────────┼──────────────────────────────────────┤
│ Valid Minimum │ Valid Parameters │       SOME Parameters at limit       │
├───────────────┴──────────────────┼──────────────────────────────────────┤
│ Below EDM threshold (goal x 10)  │           Below call limit           │
├───────────────┬──────────────────┼───────────┬─────────────┬────────────┤
│  Covariance   │     Hesse ok     │ Accurate  │  Pos. def.  │ Not forced │
└───────────────┴──────────────────┴───────────┴─────────────┴────────────┘
┌───┬──────┬───────────┬───────────┬────────────┬────────────┬─────────┬─────────┬───────┐
│   │ Name │   Value   │ Hesse Err │ Minos Err- │ Minos Err+ │ Limit-  │ Limit+  │ Fixed │
├───┼──────┼───────────┼───────────┼────────────┼────────────┼─────────┼─────────┼───────┤
│ 0 │ x0   │     1     │     7     │            │            │    0    │   10    │       │
│ 1 │ x1   │ 912.60e-3 │  0.33e-3  │            │            │  1e-10  │   10    │       │
└───┴──────┴───────────┴───────────┴────────────┴────────────┴─────────┴─────────┴───────┘
┌────┬───────────────────┐
│    │       x0       x1 │
├────┼───────────────────┤
│ x0 │       34        0 │
│ x1 │        0 1.11e-07 │
└────┴───────────────────┘
success: True

PyTorch

fit params: [0.9731879234313965, 0.9999999403953552]
┌──────────────────────────────────┬──────────────────────────────────────┐
│ FCN = 17.09                      │              Nfcn = 56               │
│ EDM = 0.000389 (Goal: 0.0002)    │              Ngrad = 1               │
├───────────────┬──────────────────┼──────────────────────────────────────┤
│ Valid Minimum │ Valid Parameters │        No Parameters at limit        │
├───────────────┴──────────────────┼──────────────────────────────────────┤
│ Below EDM threshold (goal x 10)  │           Below call limit           │
├───────────────┬──────────────────┼───────────┬─────────────┬────────────┤
│  Covariance   │     Hesse ok     │ Accurate  │  Pos. def.  │ Not forced │
└───────────────┴──────────────────┴───────────┴─────────────┴────────────┘
┌───┬──────┬───────────┬───────────┬────────────┬────────────┬─────────┬─────────┬───────┐
│   │ Name │   Value   │ Hesse Err │ Minos Err- │ Minos Err+ │ Limit-  │ Limit+  │ Fixed │
├───┼──────┼───────────┼───────────┼────────────┼────────────┼─────────┼─────────┼───────┤
│ 0 │ x0   │ 973.19e-3 │  0.29e-3  │            │            │    0    │   10    │       │
│ 1 │ x1   │1000.00e-3 │  0.24e-3  │            │            │  1e-10  │   10    │       │
└───┴──────┴───────────┴───────────┴────────────┴────────────┴─────────┴─────────┴───────┘
┌────┬───────────────────┐
│    │       x0       x1 │
├────┼───────────────────┤
│ x0 │ 8.61e-08 5.75e-08 │
│ x1 │ 5.75e-08 5.77e-08 │
└────┴───────────────────┘
success: True

JAX

WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
fit params: [0.5003563165664673, 0.9998618364334106]
┌──────────────────────────────────┬──────────────────────────────────────┐
│ FCN = 13.11                      │              Nfcn = 34               │
│ EDM = 1.23 (Goal: 0.0002)        │              Ngrad = 4               │
├───────────────┬──────────────────┼──────────────────────────────────────┤
│ Valid Minimum │ Valid Parameters │       SOME Parameters at limit       │
├───────────────┴──────────────────┼──────────────────────────────────────┤
│ Below EDM threshold (goal x 10)  │           Below call limit           │
├───────────────┬──────────────────┼───────────┬─────────────┬────────────┤
│  Covariance   │     Hesse ok     │ Accurate  │  Pos. def.  │ Not forced │
└───────────────┴──────────────────┴───────────┴─────────────┴────────────┘
┌───┬──────┬───────────┬───────────┬────────────┬────────────┬─────────┬─────────┬───────┐
│   │ Name │   Value   │ Hesse Err │ Minos Err- │ Minos Err+ │ Limit-  │ Limit+  │ Fixed │
├───┼──────┼───────────┼───────────┼────────────┼────────────┼─────────┼─────────┼───────┤
│ 0 │ x0   │     1     │     9     │            │            │    0    │   10    │       │
│ 1 │ x1   │     1     │     9     │            │            │  1e-10  │   10    │       │
└───┴──────┴───────────┴───────────┴────────────┴────────────┴─────────┴─────────┴───────┘
┌────┬───────────┐
│    │   x0   x1 │
├────┼───────────┤
│ x0 │  377 -188 │
│ x1 │ -188 94.2 │
└────┴───────────┘
success: True

```

but if you flip strategy to `2` TensorFlow is able to get the JAX result but PyTorch will fail.

Reply:
I am surprised that the TensorFlow result says `Not forced`, I thought this referred to forcing the covariance matrix to be positive definite, which seems to happen in the messages above. `SOME Parameters at limit` is also an indication that the result (and in particular parameter uncertainties) may be unreliable. It is also very odd to see `EDM = 1.14e+03 (Goal: 0.0002)` and still `success: True` (JAX also has a very large EDM). Weirdly enough, the PyTorch status looks reasonable but the result not at all.

Reply:
Is it normal to only get these few gradient calls? `Ngrad = 1` is surprising to me.

Reply:
I experimented some more with numpy and without gradients: the 32b version seems unable to reach the EDM goal, and lowering the tolerance does not help. `iminuit` still reports that the EDM is below threshold, which I do not understand even when taking into account [the factor 10](https://iminuit.readthedocs.io/en/stable/reference.html#iminuit.Minuit.tol) into account.

Reply:
Thanks for taking a look at this @alexander-held. I'm afraid I don't fully understand this behavior yet either, which might take some time.